File: ruby/050.rb
# @param {Float} x
# @param {Integer} n
# @return {Float}
def my_pow(x, n)
  return 1 if n == 0
  return 1 / my_pow(x, -n) if n < 0

  if n % 2 == 0
    value = my_pow(x, n / 2)
    value * value
  else
    value = my_pow(x, (n - 1) / 2)
    x * value * value
  end
end

File: javascript/050.js
/**
 * @param {number} x
 * @param {number} n
 * @return {number}
 */
var myPow = function(x, n) {
  if (n === 0) return 1;
  if (n === 1) return x;
  if (n < 0) return 1 / myPow(x, -n);
  
  if (n % 2 === 0) {
      var half = myPow(x, n / 2);
      return half * half;
  } else {
      var half = myPow(x, (n - 1) / 2);
      return x * half * half;
  }
};